Add a fuzzy search mode to the `search` operator

yaisog opened the pull request:
I picked up the idea in #7325 to make use of the already included diff-match-patch library's ability to do fuzzy searching.

The search operator gets a new flag `fuzzy` to make use of DMP. This can be combined with the `casesensitive` flag (which wouldn't make much sense in a fuzzy search, though). Fuzzy search is always in `literal` mode (not `words` or `some`).

The default match threshold is 0.5, but can be modified by setting a different value in `$:/config/FuzzySearchThreshold`.

The results are sorted by match score, with better matches in front, because there is no opportunity to sort post-search.

Right now, the fuzzy code is in the `search` function in wiki.js, because this caused the least duplication of code (for building the list of fields, etc.). It might be argued that this should go into the search operator code or even a new operator `fuzzy`. This would result in more duplicated code, though.

For a test run, open the _Filter_ tab in `$:/AdvancedSearch` and type something like `[all[shadows+tiddlers]search:title:fuzzy[advncd]]`.

- [ ] Check if {{$:/config/FuzzySearchThreshold}} is between 0 and 1.

Jermolene commented:
Thanks @yaisog this is an experiment.

I did notice some unexpected results. For example this yields 218 matches where it is not obvious how they are related to the search term:

```
[all[tiddlers]search:title:fuzzy[terible]]
```

It is very satisfying to be able to reuse the existing code, but I do wonder if a satisfying solution to fuzzy search might not need a more sophisticated, language aware algorithm. There are purpose built libraries like [Fuze](https://www.fusejs.io/examples.html). It looks like the version we would use is 18KB, which would probably be too large for the core, but would make a very useful plugin.

Also worth linking back to the earlier PR #7918

yaisog commented:
Hi @Jermolene, thank you for your feedback.
I think when the threshold is comparably large, there will be a lot of confusing matches. With the 0.5 used here, basically half the letters can be different in the match and it might be difficult to recognize it as such to the user. You could try reducing the threshold as described above to e.g. 0.25 – there would be no more matches to your query (as there are no corresponding titles). A threshold of 0.25 would probably work for one or maybe two missing, spurious or misspelled letters per search term, depending on its length, and might thus do just what is intended. Maybe 0.25 should be the default.

Fuze.js seems to also use the Bitap algorithm that DMP uses, I would not expect significantly better results.

Language-aware solutions that would also find "horrible" when searching for "terrible" would be awesome, but would require a thesaurus. @CrossEye proposed other metrics in #7917, but I haven't checked if there are premade libraries for that. My goal was simply to demonstrate that we could leverage the functionality that is already contained in the core to expose a fuzzy search feature.

PS: "Terrible" is the word that comes to your mind when looking at this?

yaisog commented:
I'm closing this in favor a making one or two dedicated plugins instead of messing with the core.

Jermolene commented:
Hi @yaisog I think moving this functionality to a plugin makes a lot of sense. Your prototype is very promising, but we're still at an early stage of figuring out how best to surface a UI for this. I would like to have something like this in the core once we've perfected the UI.

Just as a side thought, I do wonder whether in our case it might make sense to directly expose the threshold in the UI, perhaps by dragging a slider. It would be unusual for a conventional website to expose that level of internal workings, but in our case we've got users who value hackability, and it seems plausible that users will need to experiment with different threshold values in order to get good results with any particular corpus.

> PS: "Terrible" is the word that comes to your mind when looking at this?

Ouch no offence intended, I've edited my post.

Jermolene commented:
Hi @yaisog I meant to ask whether there are any additional core hooks/extension points that would help?

yaisog commented:
> we're still at an early stage of figuring out how best to surface a UI for this

Hi @Jermolene, do you mean as the default search on tiddlywiki.com? From my point of view, this is just another search mode like `some` or `literal` that will yield a different set of results. It would be up to the (advanced) user to make use of it in his/her templates.

The neat thing about making this a plugin is that it would give me a place to put a slider for the threshold: a plugin settings tab. Having the plugin tiddler and a tiddler with search results open next to each other would give instant feedback on the threshold effects. But I think I'd rather make this a second parameter to a new `fuzzy` filter operator.

I do wonder how I could integrate a better string similarity metric like Dice's coefficient, as proposed by @CrossEye, to score the matches that d-m-p finds (since `levenshtein` is not really how human brains score similarity), especially if the pattern is just a substring of a longer text. In this case, a subsequent Dice evaluation by another filter operator wouldn't know what to compare to what else. So it would have to be done within the search operation, where d-m-p at least returns the position of the fuzzy match so I can get the substring for similarity scoring. These scores could only be used for sorting, though, and would not be accessible to the user.

> I meant to ask whether there are any additional core hooks/extension points that would help?
I guess, I could use much of the `search` function in wiki.js if there was a way to specify a different search function to use instead of `searchTiddler` in the iterator. Otherwise, I'd copy most of the code into the plugin and modify `searchTiddler` as needed.

Jermolene commented:
> Hi @Jermolene, do you mean as the default search on tiddlywiki.com?

Yes. If we were to add this to the core then I would want to be using it to provide end user benefits. I was thinking of the discussions in #7918 about usability which suggest that we need to go a bit further to arrive at a universally useful presentation of the new functionality.

> From my point of view, this is just another search mode like `some` or `literal` that will yield a different set of results. It would be up to the (advanced) user to make use of it in his/her templates.

We tend not to have unused mechanisms in the core. For example when we introduced the DMP library we used it in the import dialogue and the internals plugin.

> The neat thing about making this a plugin is that it would give me a place to put a slider for the threshold: a plugin settings tab. Having the plugin tiddler and a tiddler with search results open next to each other would give instant feedback on the threshold effects. But I think I'd rather make this a second parameter to a new `fuzzy` filter operator.

I was thinking of making the slider be visible next to the search box (perhaps behind a disclosure triangle) so that users can dynamically see the results of changing it.

> > I meant to ask whether there are any additional core hooks/extension points that would help?
> 
> I guess, I could use much of the `search` function in wiki.js if there was a way to specify a different search function to use instead of `searchTiddler` in the iterator. Otherwise, I'd copy most of the code into the plugin and modify `searchTiddler` as needed.

Those sound like worthwhile extensions and would be worth doing, would you like to propose a PR?

yaisog commented:
> Those sound like worthwhile extensions and would be worth doing, would you like to propose a PR?

Hi @Jermolene, I took a closer look at what would need to change in wiki.search() to make it ready for fuzzy search with d-m-p. There is probably not so much that can be reused. `searchTiddler` is doing most of the work and I don't think there is a safe way to just plug in a new metric into it, so it would need to be rewritten anyway. For sorting (and possibly further thresholding), we'd need to track the match location that is returned by d-m-p to calculate the (Dice) string similarity, so there would be numerous other places to modify and/or multiple new properties in the `options` object. It's probably cleanest to create a new function and duplicate only what's needed (e.g. field handling).

On another tangent, we might reuse some of the existing logic to implement `literal` (as implemented here), `words` and `some` search modes, if it makes sense, though I haven't thought about how to score the latter ones.

CrossEye commented:
@Jermolene 

> I do wonder whether in our case it might make sense to directly expose the threshold in the UI, perhaps by dragging a slider. 

There seem to be two very distinct modes for fuzzy matching.  I would often like it when searching for a short string within tiddler titles (e.g. finding "**K**atherine Johnson" when searching for "**C**atherine".)  @Yaisog has discussed searching longer blocks such as `text` properties.  For the latter case, a scoring threshold makes a lot of sense.  For the former, though, much more likely would be to select the top `n` matches ordered by that score, regardless of the actual score.  But that `n` could well be configurable in a similar manner.

yaisog commented:
I'm gonna reopen this PR to demo the plugin functionality, as this works better on the preview site with a lot of tiddlers.

yaisog commented:
I uploaded a new self-contained version with all the code in the filter operator file. As that makes it quite large, this might not be optimal. There are too many differences to be able to make use of the core search functionality, though.

Searching is still done with diff-match-patch and the Levenshtein metric, but results are now scored and sorted with Dice's string similarity coefficient, so the first results should be the most relevant (unless there are only irrelevant results from the search).

The optimal value of the threshold can be very situation-dependent, i.e. if there are any useful results or not. I agree with @Jermolene's idea that an easy method to adjust the threshold, like a slider, should be part of the UI if fuzzy search is used.

Jermolene commented:
Thank you @yaisog I am glad to see this. The new preview does seem more effective.

I did notice a surprising difference between these two searches. The first returns what one might expect, a list of tiddlers with operator in the title. The second just returns a single result;

```
[all[tiddlers]fuzzy:title:some[opertr]]
[all[tiddlers]fuzzy:title:some[oprtr]]
```

yaisog commented:
Hi @Jermolene, thank you for trying it out. I believe the effect that you observed is directly connected to the threshold mechanic. The shorter search term probably allows for fewer errors/insertions/deletions with the threshold being a relative quantity, therefore "operator" no longer matches. I observed similar behavior in my tests. That's why I think that a slider or something similar to allow for easy access to the threshold setting would be beneficial.

If I get around to it, I'll add a _fuzzy_ tab for $:/AdvancedSearch to this PR, that provides a UI to play around with the settings a bit.

pmario commented:
@yaisog  `[all[tiddlers]fuzzy::some[opretr]]` creates an RSOD

yaisog commented:
Hi @pmario, thank you for finding this. It should be fixed now.

pmario commented on core/modules/filters/fuzzy.js:
```diff
@@ -194,8 +194,9 @@ Export our filter function
+		dmpObject = new dmp.diff_match_patch(),
+		threshold = Number(options.wiki.getTiddlerText("$:/config/FuzzySearchThreshold","0.4"));
+	dmpObject.Match_Threshold = Number.isNaN(threshold) ? 0.4 : threshold;
 	dmpObject.Match_Distance = 100000; // Large value to allow matches anywhere
```
For number parsing we do have utility functions. eg:

```
threshold = options.wiki.getTiddlerText("$:/config/FuzzySearchThreshold", FUZZY_THRESHOLD);
dmpObject.Match_Threshold = $tw.utils.parseNumber(threshold) || FUZZY_THRESHOLD;
```

See: https://github.com/TiddlyWiki/TiddlyWiki5/blob/master/core/modules/utils/utils.js#L973

yaisog replied:
Good points, thank you. Will address these.

pmario commented on core/modules/filters/fuzzy.js:
```diff
@@ -0,0 +1,237 @@
+function fuzzySearch(text,options) {
+	options = options || {};
+	var self = this,
+		t,
```
IMO this one should be moved down to the usage

pmario commented on core/modules/filters/fuzzy.js:
```diff
@@ -0,0 +1,237 @@
+	});
+	// Remove any of the results we have to exclude
+	if(options.exclude) {
+		for(t=0; t<options.exclude.length; t++) {
```
IMO should be `for(var t=0;)` ... Which is done several times in this code already.

yaisog replied:
Again, taken verbatim from search(). It was not my intention to correct existing core code. I will address this in the next version, though.

pmario replied:
IMO this is a plugin and the core has room for improvement for constructions like this one.

pmario commented on core/modules/filters/fuzzy.js:
```diff
@@ -0,0 +1,237 @@
+	if(fields.length === 0 && !options.excludeField) {
+		fields.push("title");
+		fields.push("tags");
+		fields.push("text");
```
Are we sure, to search title, tags and text by default?

If I do search at TW-com for `[all[tiddlers]fuzzy::some[advncd serch]]` on my computer it needs close to 1 second, until something will be shown.

yaisog replied:
This is taken verbatim from the regular search, trying to be as consistent as possible. Decisions on sensible defaults were not on my agenda. The preset on which fields to search could also be set in the search tiddlers.

pmario replied:
ok

yaisog replied:
Also, such things would be changed with the search template. You could always just change the search result template on the site to do a title search only (or initially).

pmario commented on core/modules/filters/fuzzy.js:
```diff
@@ -0,0 +1,237 @@
+	}
+
+	// Dice coefficient for string similarity
+	function diceCoefficient(a, b) {
```
IMO the helper functions should be moved out of the `function fuzzySearch()` function, to make the code simpler to read

yaisog replied:
Again, I tried to be consistent with search() and keep the helpers defined inside the scope in which they are used.

pmario commented on core/modules/filters/fuzzy.js:
```diff
@@ -0,0 +1,237 @@
+							str[s] = str[s].toLowerCase();
+							const matchPosition = dmpObject.match_main(str[s], terms[termIndex], 0);
+							if(matchPosition !== -1) {
+								const score = scoreMatch(terms[termIndex], str[s], matchPosition);
```
IMO `score` and `matchPosition` variable definition should be moved above the outer `for`, and it should be `var`. IMO const does not make sense here.

pmario commented on core/modules/filters/fuzzy.js:
```diff
@@ -0,0 +1,237 @@
+	// Function to check a given tiddler for the search term
+	function searchTiddler(title) {
+		var tiddler = $tw.wiki.getTiddler(title);
+		if(!tiddler) {
```
If there is no tiddler, we should return early and not create a "fake tiddler"

yaisog replied:
Again, I tried to be consistent with search() and keep the helpers defined inside the scope in which they are used.

yaisog replied:
I think this is there to be able to search within any input string, since the list of input "tiddlers" to a search operator can be any text, not just real tiddlers.

yaisog commented:
I added a _Fuzzy_ tab to the _Standard_ search in `$:/AdvancedSearch` with a slider for the threshold and a mode selector (`words`, `some` and `literal`) to make it easier to play around with. Since this respects `$:/config/Search/MinLength`, it can avoid the very slow update for 1-letter search terms.

Jermolene commented:
Hi @yaisog terrific, thank you, it's very helpful to try it out properly integrated.

yaisog commented:
Currently, `fuzzy` does not use the matchLocation within the searched string for scoring (i.e. scoring only against a substring found at that location), but scores against the whole searched string. This automatically will give a higher score to short matches due to the higher similarity to strings of similar length.

I believe that this opens up the possibility to excise the Dice similarity scoring algorithm into a separate filter operator that could be used subsequent to search to sort the results, but also for general string comparison purposes. The drawback would be that the standard sorted title,text,tags search as it is implemented now would become somewhat more complicated:
```
\function .score(target) [get[text]similarity<target>] =[get[title]similarity<target>] =[get[tags]similarity<target>] +[sum[]]

[fuzzy<searchTerm>] :sort[.score<searchTerm>]
```
or something similar.
This is not terribly flexible with regard to the field list or multiple search terms in `some` or `words` modes, though, so I don't know if this is something worth exploring. The `.score` function is probably going to end up being nested `:reduce` filter runs iterating over a list of fields and a list of search terms...
At least for a `literal` title search it will remain somewhat concise:
```
[fuzzy:title:literal<searchTerm>] :sort[similarity<searchTerm>]
```

The other alternative would be to include the whole fields and flags logic of `fuzzy` also into the similarity operator and use it like
```
[similarity:title,tags:some<targetTerm>]
```
with the same defaults of `title,text,tags` and `words`. There would be duplicated code, but the WikiText usage would be _much_ simpler...

Something to ponder, I guess. Any thoughts, @Jermolene, @pmario?

pmario commented:
@yaisog ... in 2022 I did envision a "scoring system" that's based on several parameters. See: https://talk.tiddlywiki.org/t/can-we-make-standard-search-a-little-smarter/3133/12?u=pmario

It would allow to calculate a score, which could be used for sorting, based on that score. You may have a closer look.
The results could be shown in 1 list, instead of 2 as we have it at the moment. 

-- Just an idea.

yaisog commented:
Hi @pmario, I looked at the discussion about your scoring system. I think this is something that can be implemented completely in WikiText, so I wouldn't want to make a filter operator that does that. Especially since all the weights and such should be easily configurable to account for others' opinions.

The similarity operator that I envisioned above could not easily be implemented in WikiText, if at all. So there would be a real justification for its existence. It could be another part of your search result sorting algorithm, though, just like `fuzzy` would enhance the actual search (before sorting) by also returning near-matches that might nevertheless score high.

yaisog commented:
I added two additional filter operators for testing: `fuzzy2` which will only search without sorting, and `dice` which will calculate the Dice coefficient for sorting.
For comparison I also added another search result tab, _Fuzzy2_, that uses these new operators instead of the original `fuzzy`.
It looks like the sort order has some differences between the two. To find the reason will probably mean some debugging and digging. For that I will temporarily also include the `debug-log` filter operator in this PR.

yaisog commented:
> Is it possible to fuzzy search CJK characters?

DeepWiki hat this to say about the diff-match-patch library that TW uses:
> The library is designed to work with any text content. While there aren't explicit tests for CJK characters in the test suites, there's no code that would prevent CJK characters from working correctly. The algorithm operates on Unicode strings in all language implementations.

I cannot say more, but it looks promising. Let me know if it works, should you test it.
